storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,776 @@
+            }
+            TopoProfileAction o = (TopoProfileAction) other;
+            return topoId.equals(o.topoId) && request.equals(o.request);
+        }
+        
+        @Override
+        public String toString() {
+            return "{ " + topoId + ": " + request + " }";
+        }
+    }
+    
+    static boolean equivalent(LocalAssignment a, LocalAssignment b) {
+        if (a == null && b == null) {
+            return true;
+        }
+        if (a != null && b != null) {
+            if (a.get_topology_id().equals(b.get_topology_id())) {
+                Set<ExecutorInfo> aexec = new HashSet<>(a.get_executors());
+                Set<ExecutorInfo> bexec = new HashSet<>(b.get_executors());
+                if (aexec.equals(bexec)) {
+                    boolean aHasResources = a.is_set_resources();
+                    boolean bHasResources = b.is_set_resources();
+                    if (!aHasResources && !bHasResources) {
+                        return true;
+                    }
+                    if (aHasResources && bHasResources) {
+                        if (a.get_resources().equals(b.get_resources())) {
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+        return false;
+    }
+    
+    static DynamicState stateMachineStep(DynamicState dynamicState, StaticState staticState) throws Exception {
+        LOG.debug("STATE {}", dynamicState.state);
+        switch (dynamicState.state) {
+            case EMPTY:
+                return handleEmpty(dynamicState, staticState);
+            case RUNNING:
+                return handleRunning(dynamicState, staticState);
+            case WAITING_FOR_WORKER_START:
+                return handleWaitingForWorkerStart(dynamicState, staticState);
+            case KILL_AND_RELAUNCH:
+                return handleKillAndRelaunch(dynamicState, staticState);
+            case KILL:
+                return handleKill(dynamicState, staticState);
+            case WAITING_FOR_BASIC_LOCALIZATION:
+                return handleWaitingForBasicLocalization(dynamicState, staticState);
+            case WAITING_FOR_BLOB_LOCALIZATION:
+                return handleWaitingForBlobLocalization(dynamicState, staticState);
+            default:
+                throw new IllegalStateException("Code not ready to handle a state of "+dynamicState.state);
+        }
+    }
+    
+    /**
+     * Prepare for a new assignment my downloading new required blobs, or going to empty if there is nothing to download.
+     * PRECONDITION: The slot should be empty
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws IOException on any error
+     */
+    static DynamicState prepareForNewAssignmentOnEmptySlot(DynamicState dynamicState, StaticState staticState) throws IOException {
+        assert(dynamicState.container == null);
```

Comment:
The state itself is not necessarily empty.  It could be that we just killed all of the workers.  I'll rename the function to avoid confusion.